lib/mix/lib/mix/dep/converger.ex

```diff
@@ -84,6 +84,9 @@ defmodule Mix.Dep.Converger do
     use_remote? = !!remote and Enum.any?(deps, &remote.remote?/1)
 
     if not diverged? and use_remote? do
+      # Make sure there are no cycles before calling remote converge
+      topsort(deps)
```

Comment:
@ericmj it is your call if you think we should do `deps = topsort(deps)` here or not given our conversation. Please approve or reject accordingly. :)